core/API/ResponseBuilder.php

```diff
@@ -196,18 +213,43 @@ private function handleArray($array)
                 $columnDelete = new ColumnDelete(new DataTable(), $hideColumns, $showColumns);
                 $array = $columnDelete->filter($array);
             }
-        } else if (is_numeric($firstKey)) {
-            $limit  = Common::getRequestVar('filter_limit', -1, 'integer', $this->request);
-            $offset = Common::getRequestVar('filter_offset', '0', 'integer', $this->request);
-
-            if (-1 !== $limit) {
-                $array = array_slice($array, $offset, $limit);
-            }
         }
 
         return $this->apiRenderer->renderArray($array);
     }
 
+    private function shouldApplyLimitOnArray($limit, $offset)
+    {
+        if ($limit === -1) {
+            // all fields are requested
+            return false;
+        }
+
+        if ($offset > 0) {
+            // an offset is specified, we have to apply the limit
+            return true;
+        }
+
+        // "api_datatable_default_limit" is set by API\Controller if no filter_limit is specified by the user.
+        // it holds the number of the configured default limit.
+        $limitSetBySystem = Common::getRequestVar('api_datatable_default_limit', -2, 'integer', $this->request);
```

Comment:
See comments. This is how we identify the default limit was applied in `plugins/API/Controller.php`. Didn't really find a better solution that works in all cases and is backwards compatible. Basically said, if an API returns an array we do not want to apply a set default limit by the system.  

Another solution that I tried working on was to only set the default `filter_limit` in `plugins/API/Controller.php` only if we request a report but this is getting complicated as soon as eg `API.getBulkRequest` is requested etc.